src/VisualStudio/Core/Def/Implementation/VisualStudioWorkspaceContextService.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Composition;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    internal sealed class VisualStudioWorkspaceContextService : IWorkspaceContextService
+    {
+        /// <summary>
+        /// Guid for UI context set by liveshare upon joining as a client to a session.
+        /// </summary>
+        private static readonly Guid s_sessionJoinedUIContextGuid = Guid.Parse("c6f0e3cb-a3c3-49bd-bad2-7aad8690c15b");
+        private readonly UIContext _sessionJoinedUIContext;
+
+        public VisualStudioWorkspaceContextService()
+        {
+            _sessionJoinedUIContext = UIContext.FromUIContextGuid(s_sessionJoinedUIContextGuid);
+        }
+
+        public bool IsCloudEnvironmentClient() => _sessionJoinedUIContext.IsActive;
+    }
+
+    [ExportWorkspaceServiceFactory(typeof(IWorkspaceContextService), ServiceLayer.Host), Shared]
+    internal sealed class VisualStudioWorkspaceContextServiceFactory : IWorkspaceServiceFactory
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioWorkspaceContextServiceFactory()
+        {
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            if (workspaceServices.Workspace.Kind == WorkspaceKind.AnyCodeRoslynWorkspace)
```

Comment:
So if we're only targeting the AnyCode workspace, then why not just export the service for that kind directly?

Reply:
Also, we use that kind for both local AnyCode _and_ the remote stuff? Should we be doing something else in that case?

Reply:
@jasonmalinowski it's a misleading name actually - this is what liveshare called their client side workspace and the name was just kept when we moved it.  So it doesn't apply to the local any code case.  I should probably rename this now, I don't think there's anything depending on this Kind on the liveshare side anymore.

Reply:
Oh, so if it's only used for files added in the remote case, then why do we need to check the UIContext at all to see if we're connected? It'll only be used in remote files to start with?

Reply:
Ah do you mean check if the document is in the remote workspace?  That should work, good idea.